File: LC589.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<int> preorder(Node* root) {
        vector <int> v;

        if(root == NULL)
            return v;
            
        stack <Node*> st;

        st.push(root);

        while(!st.empty()){
            Node *t = st.top();
            st.pop();

            v.push_back(t->val);

            for(auto i = t->children.rbegin(); i < t->children.rend(); i++)
                st.push(*i);
        }

        return v;
        
    }
};

//Recursive Solution

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    void traverse(Node *root, vector<int> &v){
        if(root == NULL)
            return;
        
        v.push_back(root->val);
        for(auto i : root->children)
            traverse(i,v);
    }

    vector<int> preorder(Node* root) {
        vector <int> v;

        if(root == NULL)
            return v;
            
        traverse(root, v);

        return v;
        
    }
};

File: LC589.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        l = []

        if root == None:
            return l
        
        st = []

        st.append(root)

        while len(st) != 0:
            t = st.pop()
            l.append(t.val)

            for i in range(len(t.children)-1, -1, -1):
                st.append(t.children[i])
        
        return l



# Recursive solution
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def traverse(self, root, l):
        if root == None:
            return
        
        l.append(root.val)
        for i in root.children:
            self.traverse(i,l)
        
    def preorder(self, root: 'Node') -> List[int]:
        l = []

        if root == None:
            return l
        
        self.traverse(root, l)

        return l
        
